plugins/system/stats/stats.php

```diff
@@ -45,91 +76,269 @@ class PlgSystemStats extends JPlugin
+
+		return $renderer;
+	}
+
+	/**
+	 * Get the data that will be sent to the stats server.
+	 *
+	 * @return  array.
+	 *
+	 * @since   3.5
+	 */
+	private function getStatsData()
+	{
+		return array(
+			'unique_id'   => $this->getUniqueId(),
 			'php_version' => PHP_VERSION,
 			'db_type'     => $this->db->name,
 			'db_version'  => $this->db->getVersion(),
 			'cms_version' => JVERSION,
 			'server_os'   => php_uname('s') . ' ' . php_uname('r')
 		);
+	}
 
-		// Don't let the request take longer than 2 seconds to avoid page timeout issues
-		try
-		{
-			// Don't let the request take longer than 2 seconds to avoid page timeout issues
-			JHttpFactory::getHttp()->post($this->params->get('url', 'https://developer.joomla.org/stats/submit'), $data, null, 2);
-		}
-		catch (UnexpectedValueException $e)
+	/**
+	 * Get the unique id. Generates one if none is set.
+	 *
+	 * @return  integer
+	 *
+	 * @since   3.5
+	 */
+	private function getUniqueId()
+	{
+		if (null === $this->uniqueId)
 		{
-			// There was an error sending stats. Should we do anything?
-			JLog::add('Could not send site statistics to remote server: ' . $e->getMessage(), JLog::WARNING, 'stats');
+			$this->uniqueId = $this->params->get('unique_id', hash('sha1', JUserHelper::genRandomPassword(28) . time()));
 		}
-		catch (RuntimeException $e)
+
+		return $this->uniqueId;
+	}
+
+	/**
+	 * Check if current user is allowed to send the data
+	 *
+	 * @return  boolean
+	 *
+	 * @since   3.5
+	 */
+	private function isAllowedUser()
+	{
+		return JFactory::getUser()->authorise('core.admin');
+	}
+
+	/**
+	 * Check if the debug is enabled
+	 *
+	 * @return  boolean
+	 *
+	 * @since   3.5
+	 */
+	private function isDebugEnabled()
+	{
+		return ((int) $this->params->get('debug', 0) === 1);
+	}
+
+	/**
+	 * Check if last_run + interval > now
+	 *
+	 * @return  boolean
+	 *
+	 * @since   3.5
+	 */
+	private function isUpdateRequired()
+	{
+		$last     = (int) $this->params->get('lastrun', 0);
+		$interval = (int) $this->params->get('interval', 12);
+		$mode     = (int) $this->params->get('mode', 0);
+
+		if ($mode === static::MODE_ALLOW_NEVER)
 		{
-			// There was an error connecting to the server or in the post request
-			JLog::add('Could not connect to statistics server: ' . $e->getMessage(), JLog::WARNING, 'stats');
+			return false;
 		}
-		catch (Exception $e)
+
+		// Never updated or debug enabled
+		if (!$last || !$interval || $this->isDebugEnabled())
 		{
-			// An unexpected error in processing; don't let this failure kill the site
-			JLog::add('Unexpected error connecting to statistics server: ' . $e->getMessage(), JLog::WARNING, 'stats');
+			return true;
 		}
+
+		return (abs(time() - $last) > $interval * 3600);
+	}
+
+	/**
+	 * Check valid AJAX request
+	 *
+	 * @return  boolean
+	 *
+	 * @since   3.5
+	 */
+	private function isAjaxRequest()
+	{
+		return (isset($_SERVER['HTTP_X_REQUESTED_WITH']) && strtolower($_SERVER['HTTP_X_REQUESTED_WITH']) == 'xmlhttprequest');
```

Comment:
Please can we use JInput here to grab these variables (better sanity checks) `return strtolower($this->app->input->server->get(''HTTP_X_REQUESTED_WITH', '')) == 'xmlhttprequest';`